flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LegacyScheduler.java

```diff
@@ -649,4 +658,24 @@ private String retrieveTaskManagerLocation(ExecutionAttemptID executionAttemptID
 			.map(TaskManagerLocation::toString)
 			.orElse("Unknown location");
 	}
+
+	private static boolean isPotentiallyTaskLocalException(Throwable throwable) {
```

Comment:
This method is pretty fragile. As soon as someone adds another exception, it may break. So I would suggest to add a flag in the `CheckpointFailureReason`, something like `final boolean preFlight`, that is set in the constructor and indicates if the exception at hand was thrown before the checkpoint barrier was emitted to the job (pre-flight), or not. This way, every new exception will have to set the flag and we can have in this method simply something like `isPreFlight()` which will check at which stage the exception was thrown.